File: data/code_pairs/cpp/696.cpp
#include <vector>
#include <cmath>
#include <iostream>

void SLARRA(int n, const std::vector<float>& d, std::vector<float>& e, std::vector<float>& e2,
            float spltol, float tnrm, int& nsplit, std::vector<int>& isplit, int& info) {
    const float ZERO = 0.0f;
    float eabs, tmp1;
    info = 0;

    if (n <= 0) {
        return;
    }

    nsplit = 1;
    if (spltol < ZERO) {
        tmp1 = std::abs(spltol) * tnrm;
        for (int i = 0; i < n - 1; ++i) {
            eabs = std::abs(e[i]);
            if (eabs <= tmp1) {
                e[i] = ZERO;
                e2[i] = ZERO;
                isplit[nsplit - 1] = i + 1;
                nsplit++;
            }
        }
    } else {
        for (int i = 0; i < n - 1; ++i) {
            eabs = std::abs(e[i]);
            if (eabs <= spltol * std::sqrt(std::abs(d[i])) * std::sqrt(std::abs(d[i + 1]))) {
                e[i] = ZERO;
                e2[i] = ZERO;
                isplit[nsplit - 1] = i + 1;
                nsplit++;
            }
        }
    }
    isplit[nsplit - 1] = n;
}

File: data/code_pairs/fortran/696.f90
SUBROUTINE SLARRA(N, D, E, E2, SPLTOL, TNRM, NSPLIT, ISPLIT, INFO)
  INTEGER, INTENT(IN) :: N
  REAL, INTENT(IN) :: D(*), E(*), E2(*), SPLTOL, TNRM
  INTEGER, INTENT(OUT) :: NSPLIT, ISPLIT(*), INFO
  REAL :: ZERO = 0.0E0
  INTEGER :: I
  REAL :: EABS, TMP1

  INFO = 0
  IF (N.LE.0) THEN
     RETURN
  END IF

  NSPLIT = 1
  IF (SPLTOL.LT.ZERO) THEN
     TMP1 = ABS(SPLTOL) * TNRM
     DO I = 1, N-1
        EABS = ABS(E(I))
        IF (EABS.LE.TMP1) THEN
           E(I) = ZERO
           E2(I) = ZERO
           ISPLIT(NSPLIT) = I
           NSPLIT = NSPLIT + 1
        END IF
     END DO
  ELSE
     DO I = 1, N-1
        EABS = ABS(E(I))
        IF (EABS.LE.SPLTOL * SQRT(ABS(D(I))) * SQRT(ABS(D(I+1)))) THEN
           E(I) = ZERO
           E2(I) = ZERO
           ISPLIT(NSPLIT) = I
           NSPLIT = NSPLIT + 1
        END IF
     END DO
  END IF
  ISPLIT(NSPLIT) = N
END SUBROUTINE SLARRA